WRD Guide Version 2 (#1853)
telerik/reporting-docs: user-guide/components/SubReport/creating-nested-hierarchy-with-subreports.md

---
title: Creating Nested Hierarchy
page_title: Creating Nested Hierarchy
description: Learn how to use SubReports in the Web Report Designer to create nested hierarchy.
slug: wrd-user-guide-create-nested-hierarchy-with-subreport
tags: web, report, design, components, subreport, hierarchy, master, detail, nested
published: True
reportingArea: WRDHTML5, WRDBlazorWrapper
position: 2
---

# Creating Nested Hierarchy with SubReports

> The Web Report Designer provides an alternative approach to parameter binding for hierarchical data using the **DataSource** property of the **SubReport** item. This method is cleaner when your detail report is fully data-bound (not just a table inside it) and works well for hierarchical nested data scenarios.

Reviewer: As discussed, consider moving these two sentences to the Subreport Overview and start this article with the line "Using DataSource on the SubReport..."

> Using DataSource on the SubReport item allows you to bind the child report directly to a filtered subset of the parent's data, making the hierarchy more intuitive and reducing parameter code.

Reviewer: ```suggestion
Using a DataSource in the SubReport item allows you to bind the child report directly to a filtered subset of the parent's data, making the hierarchy more intuitive and reducing parameter code.
```

> The main benefits with this approach are:

Reviewer: ```suggestion
The main benefits of this approach are:
```

          "Price": 1470.37,
        },
        {
          "ProductID": 102,
          "ProductName": "Electronics Item 102",
          "Price": 934.91,
        },
        {
          "ProductID": 103,
          "ProductName": "Electronics Item 103",
          "Price": 334.59,
        }
      ]
    },
    {
      "CategoryID": 2,
      "CategoryName": "Home Appliances",
      "Products": [
        {
          "ProductID": 104,
          "ProductName": "Home Appliances Item 104",
          "Price": 1019.39,
        },
        {
          "ProductID": 105,
          "ProductName": "Home Appliances Item 105",
          "Price": 1374.47,
           },
        {
          "ProductID": 106,
          "ProductName": "Home Appliances Item 106",
          "Price": 1128.34,
        },
        {
          "ProductID": 107,
          "ProductName": "Home Appliances Item 107",
          "Price": 1009.11,
        },
        {
          "ProductID": 108,
          "ProductName": "Home Appliances Item 108",
          "Price": 1355.37,
          },
        {
          "ProductID": 109,
          "ProductName": "Home Appliances Item 109",
          "Price": 386.92,
        }
      ]
    }
```

> 1. Create a new report (e.g. CategoriesReport.trdp) which will be used as our main report for displaying Categories.

Reviewer: Introduce the list with the steps.
```suggestion
To design a report that represents a hierarchy:

1. Create a new report (for example, `CategoriesReport.trdp`). You will use it as the main (parent) report to display categories.
```

1. Use the [Table Wizard]({%slug web-report-designer-user-guide-components-tables%}#using-the-table-wizard) to create a table bound to the JSON source.

    <img style="border: 1px solid gray;" src="images/sub-report-default-table.png" alt="Categories Products Default Table" caption="Categories Products Default Table" /> 

1. Select the generated TextBox for the Products field (e.g. the TextBox bound to Fields.Products)

1. Navigate to the Components tab and click SubReport. Thus, the TextBox item is replaced by a SubReport item.

1. Using the searchbox, navigate to the `Bindings` section and add a new binding for the **DataSource** field and setting the **Expression** to "=Fields.Products". Thus, the SubReport item will be mapped to the Products list of the respective category record.

    <img style="border: 1px solid gray;" src="images/sub-report-datasource-binding.png" alt="DataSource Binding" caption="DataSource Binding" /> 

1. Navigate to the `Inner Report` section and select the **UriReportSource** Type. Now, we will need a separate report that will be used as a container for the products for each category.

> 1. Create another report (e.g. ProductsReport.trdp) and add two TextBox items.

Reviewer: ```suggestion
1. Create another report (`ProductsReport.trdp`) and add two TextBox items.
```

 
1. Using the searchbox, navigate to the **Value** field and bind to **=Fields.ProductName** and **=Fields.Price** respectively. Save the child report.

> 1. Go back to the CategoriesReport.trdp and set the newly created ProductsReport.trdp as **Uri** for the UriReportSource.

Reviewer: ```suggestion
1. Go back to the `CategoriesReport.trdp` report and set the newly created `ProductsReport.trdp` as **Uri** for the **UriReportSource**.
```
